app/controllers/TicketBooking.scala

```diff
@@ -0,0 +1,204 @@
+package controllers
+
+import models._
+import play.api.data.Forms._
+import play.api.data._
+import play.api.mvc._
+import play.filters.csrf.CSRF
+
+import java.time.LocalDateTime
+import javax.inject._
+import scala.util.{Failure, Success, Try}
+
+@Singleton
+class TicketBooking @Inject() (cc: MessagesControllerComponents) extends MessagesAbstractController(cc) {
+
+  private val availableScreenings: List[Screening] = DefaultScreenings.getAvailableScreenings
```

Comment:
how would you provide a database storage to the solution? How the could would change?

Reply:
For this demo I would use In-memory database (like H2) with proper plugins for play and scala - maybe with slick. Unfortunately, I did not have enough time to provide it. The database would be really simple, containing one schema with a few tables: 
(1) Reservation(Id: Int <Primary Key>, screeningId: Int, customerId: Int, chosenTicketTypes: String, totalPrice: Decimal) 
* chosenTicketTypes -> it would look like a list of strings with ticket-type names separated by commas: 
*  for example: "child,adult,student,student" -- Scala functions would make proper calculations to calculate totalPrice
--> Reservation would be pointing to two tables listed below using foreign keys:
(2) Screening(Id: Int <Primary Key>, movieId: Int, screeningRoomId: Int, startTime: Timestamp, availableSeats: String)
* availableSeats: String -> it would look like a list of (row, seat) pairs separated with ";" char: "(1,1);(1,2);(1,5);(1,6)"
* after each reservation this column would be updated in scala program using scala function that would remove
* reserved places from this string
(3) Customer(Id: Int <Primary Key>, name: String, surname: String)
--> Screening table would be pointing to two tables listed below using foreign keys:
(4) Movie(id: Int <Primary Key>, title: String, durationInMinutes: Int)
(5) ScreeningRoom(id: Int <Primary Key>, numOfRows: Int, numOfSeatsPerRow: Int)
(6) TicketType(Id: Int <Primary Key>, name: String, price: Decimal)
* Ticket Type would not be connected with other tables on the database but it would be the source for the scala classes / functions that calculate prices, make reservations etc
--
The usage would look like this:
1) When a new customer begins his session on the web page data from the database is loaded into Scala code (classes / collections) + availableSeats list has to be updated more often (easier way - for example every 5 seconds, harder and probably better way - every time any other customer made a reservation on the same screening - using some signals / reactive programming style - the app server would have to have a special object the would send signals every time a reservation is made